File: js/parser.js
// Extrae entradas "Tauro" (chasis, motor, color) a partir del texto OCR de cada página.
// Heurística: busca ocurrencias del modelo Tauro y revisa una ventana de ±6 líneas
// para encontrar las etiquetas chasis / motor / color y sus valores.
//
// Formato real de las facturas Tauro:
//   MODELO : TAURO 60V
//   COLOR  : VERDE MANZANA
//   SERIAL : HEOEA2A04SA101647 | HK250101101754
//                  ^chasis           ^motor
// El chasis siempre empieza con el prefijo HEOEA2A.

(function (global) {
  "use strict";

  const CONTEXT_LINES = 8;

  // Tolerante a errores de OCR: TAURO / TAUR0 / TAURD / con espacios sueltos.
  const TAURO_RE = /T\s*A\s*U\s*R\s*[O0D]/i;

  // Prefijo conocido del chasis (Tauro). Permite OCR: HE0EA2A, HE0EAZ4, etc.
  // Captura al menos los primeros 7 chars reconocibles.
  const CHASIS_PREFIX_RE = /\bH[EF][O0][EF][A4][Z2][A4]/i;

  // Separador entre chasis y motor en la misma línea: "|", "l", "I", "1" rodeado de espacios.
  const PIPE_RE = /\s+[|lI1]\s+/;

  // Etiquetas por campo. Se evalúa en este orden (más específica primero).
  const LABELS = {
    // "SERIAL" = línea que contiene chasis | motor juntos (formato Tauro real).
    serial: [
      /\bserial\s*(?:es)?\b/i,
      /\bserie\s*(?:s)?\b/i
    ],
    chasis: [
      /(?:serial|serie|n[oº\.]?|no\.?|n[uú]mero)\s*(?:de\s*)?chasis\b/i,
      /\bchasis\b/i,
      /\bvin\b/i
    ],
    motor: [
      /(?:serial|serie|n[oº\.]?|no\.?|n[uú]mero)\s*(?:de\s*)?motor\b/i,
      /\bmotor\b/i
    ],
    color: [
      /\bcolor\b/i
    ]
  };

  // VIN "canónico": 17 alfanuméricos sin I, O, Q.
  const VIN_RE = /\b[A-HJ-NPR-Z0-9]{17}\b/;
  // Respaldo genérico para chasis/motor: 8–20 alfanuméricos.
  const GENERIC_SERIAL_RE = /\b[A-Z0-9][A-Z0-9\-]{7,19}\b/;
  // Motor suele ser más corto.
  const MOTOR_SERIAL_RE = /\b[A-Z0-9][A-Z0-9\-]{5,19}\b/;

  function splitLines(text) {
    return String(text || "")
      .replace(/\r/g, "")
      .split("\n")
      .map(l => l.trim())
      .filter(l => l.length > 0);
  }

  function findFirstMatch(line, regexes) {
    for (const re of regexes) {
      const m = line.match(re);
      if (m) return { index: m.index, length: m[0].length, label: m[0] };
    }
    return null;
  }

  // Decide si un token alphanumerico es chasis (por prefijo) o motor.
  function classifyToken(token) {
    return CHASIS_PREFIX_RE.test(token) ? "chasis" : "motor";
  }

  // Maneja la línea "SERIAL : HEOEA2A04SA101647 | HK250101101754"
  // Devuelve { chasis, motor } o null si no coincide el patrón.
  function extractSerialLine(line, nextLine) {
    const hasLabel = findFirstMatch(line, LABELS.serial);
    if (!hasLabel) return null;

    const after = line.slice(hasLabel.index + hasLabel.length)
      .replace(/^[\s:\-.·•]+/, "");

    // Acepta que el contenido pueda estar en la línea siguiente si after está vacío.
    const source = after.length >= 8 ? after : (nextLine || "");
    if (!source) return null;

    // ¿Hay separador pipe (real u OCR)? → divide en dos partes.
    if (PIPE_RE.test(source)) {
      const parts = source.split(PIPE_RE).map(p => p.trim().toUpperCase());
      let chasis = null, motor = null;
      parts.forEach(p => {
        const tokens = p.match(GENERIC_SERIAL_RE) || [];
        tokens.forEach(t => {
          if (!chasis && classifyToken(t) === "chasis") chasis = t;
          else if (!motor) motor = t;
        });
      });
      // Si no se pudo clasificar por prefijo, primer token = chasis.
      if (!chasis && parts.length >= 2) {
        const t0 = (parts[0].match(GENERIC_SERIAL_RE) || [])[0];
        const t1 = (parts[1].match(GENERIC_SERIAL_RE) || [])[0];
        chasis = t0 || null;
        motor  = t1 || null;
      }
      if (chasis || motor) return { chasis, motor };
    }

    // Sin pipe: extrae todos los tokens y clasifica por prefijo.
    const tokens = source.toUpperCase().match(GENERIC_SERIAL_RE) || [];
    let chasis = null, motor = null;
    tokens.forEach(t => {
      if (!chasis && classifyToken(t) === "chasis") chasis = t;
      else if (!motor) motor = t;
    });
    if (chasis || motor) return { chasis, motor };
    return null;
  }

  // Busca el valor del campo a partir de una etiqueta encontrada en `line`:
  //   primero tras la etiqueta en la misma línea (p.ej. "CHASIS: ABC123...")
  //   y si no aparece, en la línea siguiente.
  function extractValue(field, line, nextLine) {
    const labelMatch = findFirstMatch(line, LABELS[field]);
    if (!labelMatch) return null;

    const after = line.slice(labelMatch.index + labelMatch.length)
      .replace(/^[\s:\-.·•]+/, "");

    const tryExtract = (source) => {
      if (!source) return null;
      if (field === "chasis") {
        const vin = source.match(VIN_RE);
        if (vin) return vin[0];
        const gen = source.toUpperCase().match(GENERIC_SERIAL_RE);
        return gen ? gen[0] : null;
      }
      if (field === "motor") {
        const gen = source.toUpperCase().match(MOTOR_SERIAL_RE);
        return gen ? gen[0] : null;
      }
      if (field === "color") {
        // 1 o 2 palabras alfabéticas, acepta tildes y "ñ".
        const m = source.match(/\b([A-Za-zÁÉÍÓÚÜÑáéíóúüñ]{3,})(?:\s+([A-Za-zÁÉÍÓÚÜÑáéíóúüñ]{3,}))?/);
        if (!m) return null;
        return (m[2] ? `${m[1]} ${m[2]}` : m[1]).toUpperCase();
      }
      return null;
    };

    return tryExtract(after) || tryExtract(nextLine);
  }

  function normalizeSerial(s) {
    return String(s || "").toUpperCase().replace(/[^A-Z0-9]/g, "");
  }

  function normalizeColor(s) {
    return String(s || "")
      .normalize("NFD").replace(/[̀-ͯ]/g, "")
      .toUpperCase().trim();
  }

  // Agrupa Tauro que parecen ser el mismo ítem dentro de una misma factura:
  // si dos ocurrencias están muy cerca y comparten al menos un campo, se fusionan.
  function dedupeWithinInvoice(entries) {
    const out = [];
    for (const e of entries) {
      const twin = out.find(o =>
        o.file === e.file && o.page === e.page &&
        (
          (o.chasisNorm && o.chasisNorm === e.chasisNorm) ||
          (o.motorNorm  && o.motorNorm  === e.motorNorm)
        )
      );
      if (twin) {
        twin.chasis = twin.chasis || e.chasis;
        twin.motor  = twin.motor  || e.motor;
        twin.color  = twin.color  || e.color;
        twin.chasisNorm = normalizeSerial(twin.chasis);
        twin.motorNorm  = normalizeSerial(twin.motor);
        twin.colorNorm  = normalizeColor(twin.color);
      } else {
        out.push(e);
      }
    }
    return out;
  }

  // `pages` es un array de strings (texto por página) y `fileName` identifica la factura.
  function extractTauroEntries(pages, fileName) {
    const entries = [];
    pages.forEach((pageText, pageIdx) => {
      const lines = splitLines(pageText);
      if (!lines.length) return;

      const hits = [];
      lines.forEach((ln, i) => { if (TAURO_RE.test(ln)) hits.push(i); });

      hits.forEach(hitIdx => {
        const start = Math.max(0, hitIdx - CONTEXT_LINES);
        const end   = Math.min(lines.length - 1, hitIdx + CONTEXT_LINES);
        let chasis = null, motor = null, color = null;

        for (let i = start; i <= end; i++) {
          const line = lines[i];
          const next = lines[i + 1] || "";

          // Formato Tauro real: SERIAL : chasis | motor en una sola línea.
          if (!chasis || !motor) {
            const sr = extractSerialLine(line, next);
            if (sr) {
              chasis = chasis || sr.chasis;
              motor  = motor  || sr.motor;
            }
          }

          // Fallback: etiquetas separadas (CHASIS: xxx / MOTOR: yyy).
          if (!chasis) chasis = extractValue("chasis", line, next);
          if (!motor)  motor  = extractValue("motor",  line, next);
          if (!color)  color  = extractValue("color",  line, next);
        }

        // Solo registramos la entrada si al menos un campo es útil.
        if (chasis || motor || color) {
          entries.push({
            file: fileName,
            page: pageIdx + 1,
            chasis: chasis ? chasis.toUpperCase() : null,
            motor:  motor  ? motor.toUpperCase()  : null,
            color:  color  ? normalizeColor(color) : null,
            chasisNorm: normalizeSerial(chasis),
            motorNorm:  normalizeSerial(motor),
            colorNorm:  normalizeColor(color)
          });
        }
      });
    });

    return dedupeWithinInvoice(entries);
  }

  global.ParserModule = {
    extractTauroEntries,
    normalizeSerial,
    normalizeColor
  };
})(window);

File: server/services/parser.py
"""
Extrae entradas de vehículos (modelo, chasis, motor, color) del texto OCR.
Soporta cualquier modelo (multi-modelo): recibe known_models como parámetro.
Porta la lógica de js/parser.js conservando las mismas heurísticas.

Formato principal de facturas:
  MODELO : TAURO 60V
  COLOR  : VERDE MANZANA
  SERIAL : HEOEA2A04SA101647 | HK250101101754
                ^chasis            ^motor
"""

import re
import unicodedata
from difflib import get_close_matches
from typing import Optional

CONTEXT_LINES = 8

# Prefijo conocido del chasis. Tolerante a errores OCR.
CHASIS_PREFIX_RE = re.compile(r'\bH[EF][O0][EF][A4][Z2][A4]', re.I)

# Separador chasis|motor: "|", "l", "I", "1" con espacios.
PIPE_RE = re.compile(r'\s+[|lI1]\s+')

# Seriales genéricos (8-20 alfanum/guiones)
GENERIC_SERIAL_RE = re.compile(r'\b[A-Z0-9][A-Z0-9\-]{7,19}\b')
MOTOR_SERIAL_RE   = re.compile(r'\b[A-Z0-9][A-Z0-9\-]{5,19}\b')
# VIN canónico 17 chars (sin I, O, Q)
VIN_RE = re.compile(r'\b[A-HJ-NPR-Z0-9]{17}\b')

# Etiquetas por campo
LABEL_SERIAL  = re.compile(r'\bserial(?:es)?\b|\bserie(?:s)?\b', re.I)
LABEL_CHASIS  = re.compile(
    r'(?:serial|serie|n[oº.]?|no\.?|n[uú]mero)\s*(?:de\s*)?chasis\b|\bchasis\b|\bvin\b', re.I)
LABEL_MOTOR   = re.compile(
    r'(?:serial|serie|n[oº.]?|no\.?|n[uú]mero)\s*(?:de\s*)?motor\b|\bmotor\b', re.I)
LABEL_COLOR   = re.compile(r'\bcolor\b', re.I)
LABEL_MODELO  = re.compile(r'\bmodelo\b', re.I)

SEPARATOR_RE  = re.compile(r'^[\s:\-.·•]+')
COLOR_WORDS   = re.compile(
    r'\b([A-Za-zÀ-ɏ]{3,})(?:\s+([A-Za-zÀ-ɏ]{3,}))?')


# ── Normalización ──────────────────────────────────────────────────────────

def normalize_serial(s: Optional[str]) -> str:
    if not s:
        return ""
    return re.sub(r'[^A-Z0-9]', '', s.upper())


def normalize_color(s: Optional[str]) -> str:
    if not s:
        return ""
    nfd = unicodedata.normalize('NFD', s)
    stripped = ''.join(c for c in nfd if unicodedata.category(c) != 'Mn')
    return stripped.upper().strip()


# ── Extracción de campos ───────────────────────────────────────────────────

def _after_label(label_re: re.Pattern, line: str) -> str:
    m = label_re.search(line)
    if not m:
        return ""
    return SEPARATOR_RE.sub('', line[m.end():])


def _extract_chasis_value(source: str) -> Optional[str]:
    vin = VIN_RE.search(source.upper())
    if vin:
        return vin.group()
    gen = GENERIC_SERIAL_RE.search(source.upper())
    return gen.group() if gen else None


def _extract_motor_value(source: str) -> Optional[str]:
    gen = MOTOR_SERIAL_RE.search(source.upper())
    return gen.group() if gen else None


def _extract_color_value(source: str) -> Optional[str]:
    m = COLOR_WORDS.search(source)
    if not m:
        return None
    result = m.group(1)
    if m.group(2):
        result += ' ' + m.group(2)
    return result.upper()


def _classify_token(token: str) -> str:
    """Devuelve 'chasis' si el token empieza con el prefijo conocido, si no 'motor'."""
    return 'chasis' if CHASIS_PREFIX_RE.search(token) else 'motor'


def _extract_serial_line(line: str, next_line: str) -> Optional[dict]:
    """Maneja: SERIAL : HEOEA2A... | HK250...  → {chasis, motor}"""
    if not LABEL_SERIAL.search(line):
        return None
    after = _after_label(LABEL_SERIAL, line)
    source = after if len(after) >= 8 else next_line
    if not source:
        return None

    chasis = motor = None

    if PIPE_RE.search(source):
        parts = [p.strip().upper() for p in PIPE_RE.split(source)]
        for part in parts:
            tokens = GENERIC_SERIAL_RE.findall(part)
            for t in tokens:
                if not chasis and _classify_token(t) == 'chasis':
                    chasis = t
                elif not motor:
                    motor = t
        if not chasis and len(parts) >= 2:
            t0 = GENERIC_SERIAL_RE.search(parts[0])
            t1 = GENERIC_SERIAL_RE.search(parts[1])
            chasis = t0.group() if t0 else None
            motor  = t1.group() if t1 else None
    else:
        tokens = GENERIC_SERIAL_RE.findall(source.upper())
        for t in tokens:
            if not chasis and _classify_token(t) == 'chasis':
                chasis = t
            elif not motor:
                motor = t

    if chasis or motor:
        return {'chasis': chasis, 'motor': motor}
    return None


# ── Localización de modelos ────────────────────────────────────────────────

def _find_model_hits(lines: list[str], known_models: list[str]) -> list[int]:
    """Devuelve índices de líneas donde aparece algún modelo conocido."""
    hits = []
    for i, line in enumerate(lines):
        upper = line.upper()
        for model in known_models:
            mu = model.upper()
            # Coincidencia directa
            if mu in upper:
                hits.append(i)
                break
            # Fuzzy por tokens (tolera errores OCR)
            line_tokens = upper.split()
            model_tokens = mu.split()
            if all(get_close_matches(mt, line_tokens, n=1, cutoff=0.82)
                   for mt in model_tokens):
                hits.append(i)
                break
    return hits


def _extract_model_name(line: str, known_models: list[str]) -> Optional[str]:
    """Obtiene el modelo reconocido en la línea."""
    upper = line.upper()
    for model in known_models:
        mu = model.upper()
        if mu in upper:
            return model
    return None


# ── Extracción principal ───────────────────────────────────────────────────

def _dedupe_within_invoice(entries: list[dict]) -> list[dict]:
    out: list[dict] = []
    for e in entries:
        twin = next(
            (o for o in out
             if o['file'] == e['file'] and o['page'] == e['page'] and (
                 (o['chasis_norm'] and o['chasis_norm'] == e['chasis_norm']) or
                 (o['motor_norm']  and o['motor_norm']  == e['motor_norm'])
             )),
            None
        )
        if twin:
            for field in ('chasis', 'motor', 'color', 'modelo'):
                twin[field] = twin[field] or e[field]
            twin['chasis_norm'] = normalize_serial(twin['chasis'])
            twin['motor_norm']  = normalize_serial(twin['motor'])
            twin['color_norm']  = normalize_color(twin['color'])
        else:
            out.append(e)
    return out


def extract_entries(pages: list[str], filename: str,
                    known_models: list[str]) -> list[dict]:
    """
    Extrae todas las entradas de vehículos del texto OCR de un PDF.
    known_models: lista de modelos a buscar; si vacía, busca cualquier
                  línea con etiqueta MODELO:.
    """
    entries: list[dict] = []

    for page_idx, page_text in enumerate(pages):
        lines = [l.strip() for l in page_text.replace('\r', '').split('\n')
                 if l.strip()]
        if not lines:
            continue

        if known_models:
            hits = _find_model_hits(lines, known_models)
        else:
            # Modo "ALL sin inventario": usa etiqueta MODELO como pivot
            hits = [i for i, l in enumerate(lines) if LABEL_MODELO.search(l)]

        for hit_idx in hits:
            start = max(0, hit_idx - CONTEXT_LINES)
            end   = min(len(lines) - 1, hit_idx + CONTEXT_LINES)

            chasis = motor = color = modelo = None

            # Intentar extraer modelo de la línea que disparó el hit
            if known_models:
                modelo = _extract_model_name(lines[hit_idx], known_models)
            else:
                after = _after_label(LABEL_MODELO, lines[hit_idx])
                modelo = after.strip() or None

            for i in range(start, end + 1):
                line = lines[i]
                next_line = lines[i + 1] if i + 1 < len(lines) else ""

                # Formato principal: SERIAL : chasis | motor
                if not chasis or not motor:
                    sr = _extract_serial_line(line, next_line)
                    if sr:
                        chasis = chasis or sr.get('chasis')
                        motor  = motor  or sr.get('motor')

                # Fallback: etiquetas separadas
                if not chasis:
                    after = _after_label(LABEL_CHASIS, line)
                    if after:
                        chasis = (_extract_chasis_value(after)
                                  or _extract_chasis_value(next_line))
                if not motor:
                    after = _after_label(LABEL_MOTOR, line)
                    if after:
                        motor = (_extract_motor_value(after)
                                 or _extract_motor_value(next_line))
                if not color:
                    after = _after_label(LABEL_COLOR, line)
                    if after:
                        color = (_extract_color_value(after)
                                 or _extract_color_value(next_line))
                if not modelo:
                    after = _after_label(LABEL_MODELO, line)
                    if after:
                        modelo = after.strip() or None

            if chasis or motor or color:
                entries.append({
                    'file': filename,
                    'page': page_idx + 1,
                    'modelo': modelo,
                    'chasis': chasis.upper() if chasis else None,
                    'motor':  motor.upper()  if motor  else None,
                    'color':  normalize_color(color) if color else None,
                    'chasis_norm': normalize_serial(chasis),
                    'motor_norm':  normalize_serial(motor),
                    'color_norm':  normalize_color(color),
                })

    return _dedupe_within_invoice(entries)
